Updating Project/Namespace annotations for displayName & description

stevekuznetsov opened the pull request:
Prepending 'openshift.io/' to 'display-name'  & ' description' annotation fields for namespacing. 

Changing the projectRequest struct to stop mirroring the project struct.

@derekwaynecarr 
#2169

liggitt commented on assets/app/views/createProject.html:
```diff
@@ -34,19 +34,19 @@
           <label for="description">Description</label>
           <textarea class="form-control input-lg"
-              name="displayName"
+              name="openshift.io/displayName"
```
this should actually have been `description`

liggitt commented on assets/app/views/createProject.html:
```diff
@@ -34,19 +34,19 @@
 
         <div class="form-group">
-          <label for="displayName">Display Name</label>
+          <label for="openshift.io/displayName">Display Name</label>
```
don't need to change the html field names, only where we actually read the annotation off the API object

liggitt commented:
I think the pattern we're trying to move to is hyphenated annotation keys rather than camel-case

liggitt commented:
Also need to search for `.displayName`, `.description`, `displayName:`, and `description:` in javascript and html to catch places we were using dot notation to reference, or JSON notation to set those annotations.

Anywhere you find us reading annotations directly should switch to using the `annotation` filter (or the `displayName` filter, which falls back to the object name if no display name annotation is present).

Example usage:
JS: `$filter('displayName')(project)`
HTML: `{{project | displayName}}`

Additional places these annotations are used:
scripts/controllers/newfromtemplate.js
scripts/controllers/createProject.js

liggitt commented on pkg/cmd/admin/project/new_project.go:
```diff
@@ -90,8 +90,8 @@ func (o *NewProjectOptions) Run() error {
-	project.Annotations["description"] = o.Description
-	project.Annotations["displayName"] = o.DisplayName
+	project.Annotations["openshift.io/description"] = o.Description
+	project.Annotations["openshift.io/displayName"] = o.DisplayName
```
add constants for these to `project/api/types.go` and use them everywhere

liggitt commented on pkg/cmd/cli/describe/printer.go:
```diff
@@ -152,7 +152,7 @@ func PrintTemplateParameters(params []templateapi.Parameter, output io.Writer) e
 	description := ""
 	if t.Annotations != nil {
-		description = t.Annotations["description"]
+		description = t.Annotations["openshift.io/description"]
```
constant in `template/api/types.go`

liggitt replied:
actually, not sure you meant to change this... this is for Template...

stevekuznetsov replied:
Template will follow a different namespace convention? #2169 references lots of annotations for Template that need updating, what to do with those?

liggitt commented on pkg/cmd/cli/describe/projectstatus_test.go:
```diff
@@ -40,7 +40,7 @@ func TestProjectStatus(t *testing.T) {
 						Namespace: "",
 						Annotations: map[string]string{
-							"displayName": "Test",
+							"openshift.io/displayName": "Test",
```
the tests can inline the strings, rather than reference the constants, because we want tests to break if we change a constant people have coded against

liggitt commented on pkg/project/registry/projectrequest/delegated/sample_template.go:
```diff
@@ -29,8 +29,8 @@ func DefaultTemplate() *templateapi.Template {
 	project.Annotations = map[string]string{
-		"description": "${" + ProjectDescriptionParam + "}",
-		"displayName": "${" + ProjectDisplayNameParam + "}",
+		"description":              "${" + ProjectDescriptionParam + "}",
```
prefix description?

liggitt commented:
The more I think about this, the more I think display name and description should be first-class, non-namespaced annotations. We would want to pick up and use kubernetes display names and descriptions if they existed, not have "openshift display name".

stevekuznetsov commented:
Reverted changes to namespace @liggitt

derekwaynecarr commented:
@liggitt we have no pool of standard annotation key names in upstream.  in the absence of that, i would think we want to prefix.

derekwaynecarr commented:
@liggitt - also upstream is likely to also prefix as kubernetes.io/display-name, if it were to have standard names.

liggitt commented:
blah

stevekuznetsov commented:
@derekwaynecarr @liggitt  So we DID want the prefix?

derekwaynecarr commented:
Prefix all things.

stevekuznetsov commented:
Fixed @derekwaynecarr @liggitt

stevekuznetsov commented:
[test]

derekwaynecarr commented:
Backend pieces LGTM, will let Jordan comment on the UI pieces.

liggitt commented on assets/app/scripts/controllers/newfromtemplate.js:
```diff
@@ -8,7 +8,7 @@
  */
 angular.module('openshiftConsole')
-  .controller('NewFromTemplateController', function ($scope, $http, $routeParams, DataService, $q, $location, TaskList, $parse, Navigate) {
+  .controller('NewFromTemplateController', function ($scope, $http, $routeParams, DataService, $q, $location, TaskList, $parse, Navigate, $filter) {
```
look up the `displayName` filter once and save it in a var:
`var displayNameFilter = $filter('displayName');`

liggitt commented on assets/app/scripts/controllers/newfromtemplate.js:
```diff
@@ -84,11 +84,11 @@ angular.module('openshiftConsole')
 
     $scope.projectDisplayName = function() {
-      return (this.project && this.project.displayName) || this.projectName;
+      return (this.project && $filter('displayName')(this.project)) || this.projectName;
```
`return displayNameFilter(this.project) || this.projectName;`

liggitt commented on assets/app/scripts/controllers/newfromtemplate.js:
```diff
@@ -84,11 +84,11 @@ angular.module('openshiftConsole')
 
     $scope.templateDisplayName = function() {
-      return (this.template.annotations && this.template.annotations.displayName) || this.template.metadata.name;
+      return (this.template.annotations && $filter('displayName')(this.template)) || this.template.metadata.name;
```
`return displayNameFilter(this.template);`

liggitt commented on pkg/project/registry/projectrequest/delegated/delegated.go:
```diff
@@ -69,8 +69,8 @@ func (r *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
 
-	if len(projectRequest.Annotations["description"]) > 0 {
-		projectDescription = projectRequest.Annotations["description"]
+	if len(projectRequest.Annotations[projectapi.ProjectDescription]) > 0 {
```
createProject.js is still sending the old annotation:
```
        DataService.create('projectRequests', null, {
          name: $scope.name,
          displayName: $scope.displayName,
          annotations: {
            description: $scope.description
          }
          ...
```

Needs to be:
```
            'openshift.io/description': $scope.description
```

liggitt replied:
Make sure you can create a project using the three methods and set the display name and description:
`osadm new-project`, `osc new-project`, and the web UI

stevekuznetsov replied:
@liggitt in createProject.js - currently why is `description` currently not an inline string? Also - should we be passing in the `displayName` as it's own parameter or moving it into the annotation list?

liggitt commented:
@bparees @derekwaynecarr do you know of project json files in other repos that need the annotation key updated?

derekwaynecarr commented:
I am not aware of other repos with project files.  

Sent from my iPhone

> On May 28, 2015, at 2:24 AM, Jordan Liggitt <notifications@github.com> wrote:
> 
> @bparees @derekwaynecarr do you know of project json files in other repos that need the annotation key updated?
> 
> —
> Reply to this email directly or view it on GitHub.
>

stevekuznetsov commented:
@liggitt Made the requested changes to ProjectRequest

stevekuznetsov commented:
[test]

liggitt commented on assets/app/scripts/controllers/createProject.js:
```diff
@@ -13,10 +13,8 @@ angular.module('openshiftConsole')
-            description: $scope.description
-          }
+          DisplayName: $scope.displayName,
+          Description: $scope.description
```
API fields always start with lowercase in their serialized form

stevekuznetsov commented:
[test]

stevekuznetsov commented:
@liggitt test-assets and test-go still run without errors on my end - anything else to do here?

liggitt commented on examples/hello-openshift/v1beta3/hello-project.json:
```diff
@@ -8,8 +8,8 @@
-      "description": "This is an example project to demonstrate OpenShift v3",
-      "displayName": "Hello OpenShift"
+      "openshift.io/description": "This is an example project to demonstrate OpenShift v3",
+      "openshift.io/splay-name": "Hello OpenShift"
```
typo: `"openshift.io/splay-name"`

liggitt commented:
one typo, LGTM otherwise

stevekuznetsov commented:
@liggitt should be fixed

liggitt commented:
[merge]

stevekuznetsov commented:
all traces of projectapi.NodeSelectorParam () should be gone by now... also removed from tests and replaced with literals to test against constants changing as per above

liggitt commented on pkg/project/api/v1/types.go:
```diff
@@ -45,4 +43,5 @@ type ProjectRequest struct {
 	kapi.TypeMeta   `json:",inline"`
 	kapi.ObjectMeta `json:"metadata,omitempty"`
 	DisplayName     string `json:"displayName,omitempty" description:"display name to apply to a project"`
+	Description     string `json:"description,omitempty"`
```
descriptions are required now

stevekuznetsov commented:
@liggitt added description for description field to v1 api

liggitt commented:
uhhh... not sure what merged. can you check and open a follow-up PR with the description if needed?

stevekuznetsov commented:
@liggitt  looks like it made it in https://github.com/openshift/origin/blob/master/pkg/project/api/v1/types.go#L46
